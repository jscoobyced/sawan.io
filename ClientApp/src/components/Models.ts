import { match } from "react-router-dom";

export interface SelectOption {
    value: number;
    text: string;
}

export interface IdParam {
    id: string;
}
