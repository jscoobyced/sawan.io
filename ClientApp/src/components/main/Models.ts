import { NavigationMenuContent } from "../../services/Models";

export interface HeaderProps {
    navigationMenuContent: NavigationMenuContent;
    signIn: (isSignedIn: boolean) => void;
    isSignedIn: boolean;
}
