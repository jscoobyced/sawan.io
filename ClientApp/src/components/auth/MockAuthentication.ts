import { IAuthentication } from "../../utils/IAuthentication";

export class MockAuthentication implements IAuthentication {
    public getAuthenticatedUser = () => {
        return {
            email: "user@example.org",
            family_name: "Smith",
            given_name: "Kevin",
            group: "Admin",
            id: "123456789",
            locale: "en_US",
            name: "Kevin Smith",
            picture: "",
            token: "",
            verified_email: true
        };
    }

    public getTokenId = (): string => {
        return "123456789";
    }

    public init = (callBack: () => void, buttonId: string) => {
        this.renderButton();
        callBack();
    }

    public isReady = (): boolean => {
        return true;
    }

    public isAuthSupported = (): boolean => {
        return true;
    }

    public signOut = (): Promise<any> => {
        return Promise.resolve();
    }

    public renderButton = () => {
        return;
    }

    public isAdmin() {
        return true;
    }
}
