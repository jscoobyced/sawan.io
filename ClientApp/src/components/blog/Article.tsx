import React from 'react';
import { NavLink } from 'react-router-dom';
import { BlogElement } from '../../services/Models';
import { DateUtil } from '../../utils/DateUtils';
import { HtmlUtils } from '../../utils/HtmlUtils';

export interface ArticleProps {
    blogElement: BlogElement;
    ellipsis?: boolean;
    backLink?: boolean;
    editLink?: boolean;
}

export class Article extends React.Component<ArticleProps> {
    public constructor(props: ArticleProps) {
        super(props);
    }

    public render() {
        const article = this.props.blogElement;
        if (!article) {
            return null;
        }
        let content = HtmlUtils.getSafeContent(article.article);
        if (this.props.ellipsis) {
            content = HtmlUtils.getEllipsis(content, 1000);
        }
        const blogDate = DateUtil.dateToString(!article.blogDate ? DateUtil.defaultDate() : article.blogDate);
        let editLink = null;
        let readMoreLink = null;
        let backLink = null;
        if (this.props.ellipsis) {
            readMoreLink = <NavLink to={'/blog/view/' + article.id} className='link-article'>Read more...</NavLink>;
        }
        if (this.props.backLink) {
            backLink = <NavLink to='/' className='btn-article'>Back</NavLink>;
        }
        if (this.props.editLink) {
            editLink = <NavLink to={'/blog/edit/' + article.id}>Edit</NavLink>;
        }

        return <div>
            <article className='blog-article'>
                <header>
                    <h1>{this.props.blogElement.articleTitle}</h1>
                    <span>{blogDate}</span>
                </header>
                <div className='article-content'>
                    <span dangerouslySetInnerHTML={{ __html: content }}></span>
                    <div className='article-links'>
                        {editLink}
                        {readMoreLink}
                    </div>
                </div>
                {backLink}
            </article>
        </div>;
    }
}
