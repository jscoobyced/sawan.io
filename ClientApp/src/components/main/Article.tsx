import React from 'react';
import { NavLink } from 'react-router-dom';
import { BlogElement } from '../../services/Models';
import { DateUtil } from '../../utils/DateUtils';
import { HtmlUtils } from '../../utils/HtmlUtils';

export interface ArticleProps {
    blogElement: BlogElement;
    ellipsis?: boolean;
    backLink?: boolean;
}

export class Article extends React.Component<ArticleProps> {
    public constructor(props: ArticleProps) {
        super(props);
    }

    public render() {
        const article = this.props.blogElement;
        if (!article) {
            return null;
        }
        let content = HtmlUtils.getSafeContent(article.article);
        if (this.props.ellipsis) {
            content = HtmlUtils.getEllipsis(content, 1000);
        }
        const blogDate = DateUtil.dateToString(article.blogDate);
        return <div>
            <article>
                <h1>{this.props.blogElement.articleTitle} - {blogDate}</h1>
                <div>
                    <span dangerouslySetInnerHTML={{ __html: content }}></span>
                    {
                        this.props.ellipsis
                            ? <NavLink to={'/blog/' + article.id} className='link-article'>Read more...</NavLink>
                            : null
                    }
                </div>
            </article>
            {
                this.props.backLink
                    ? <NavLink to='/' className='link-article'>Back</NavLink>
                    : null

            }
        </div>;
    }
}