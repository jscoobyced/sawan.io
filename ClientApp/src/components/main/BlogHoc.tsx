import React from 'react';
import { match } from 'react-router-dom';
import { ContentServiceFactory } from '../../services/ContentServiceFactory';
import { IContentService } from '../../services/IContentService';
import { BlogElement } from '../../services/Models';
import { IdParam } from '../Models';
import { Blog } from './Blog';

export interface BlogHocProps {
    match: match<IdParam>;
}

export interface BlogState {
    blogElement: BlogElement;
}

export class BlogHoc extends React.Component<BlogHocProps, BlogState> {
    private readonly contentService: IContentService;

    public constructor(props: BlogHocProps, state: BlogState) {
        super(props, state);
        this.contentService = ContentServiceFactory.GetContentService();
        this.state = {
            blogElement: {
                article: '',
                articleTitle: '',
                blogDate: new Date("2019-01-01T00:00:00+07:00"),
                id: 0
            }
        };
    }

    public componentDidMount() {
        this.contentService.getBlogElement(this.props.match.params.id)
            .then(blogElement => {
                if (blogElement) {
                    this.setState({ blogElement });
                }
            });
    }

    public render() {
        return <Blog blogElement={this.state.blogElement} />;
    }
}
