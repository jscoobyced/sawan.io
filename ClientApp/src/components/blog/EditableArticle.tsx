import React from 'react';
import { BlogElement } from '../../services/Models';
import { DateUtil } from '../../utils/DateUtils';
import { Article } from './Article';

export interface EditableArticleProps {
    blogElement: BlogElement;
    update: (article: string, articleTitle: string, articleBlogDate: Date, id: number) => Promise<boolean>;
}

export interface EditableArticleState {
    blogElement: BlogElement;
    saveText: string;
    saveDisabled: boolean;
}

export class EditableArticle extends React.Component<EditableArticleProps, EditableArticleState> {
    private readonly SAVE = 'Save';
    private readonly SAVING = 'Saving';
    private readonly SAVED = 'Saved';
    private readonly FAILED = 'Failed';
    public constructor(props: EditableArticleProps) {
        super(props);
        this.state = {
            blogElement: props.blogElement,
            saveText: this.SAVE,
            saveDisabled: false
        };
    }

    public componentDidUpdate(prevProps: EditableArticleProps) {
        const blogElement = this.props.blogElement;
        if (blogElement.article !== prevProps.blogElement.article) {
            this.setState({
                blogElement
            });
        }
    }

    public render() {
        if (!this.state.blogElement) {
            return <></>;
        }

        const { article, articleTitle } = this.state.blogElement;
        return (
            <div>
                <article
                    className='blog-article'
                >
                    <input
                        type='text'
                        id='article-title'
                        value={articleTitle}
                        onChange={this.onChangeArticleTitle}
                    />
                    <textarea
                        rows={24}
                        id='article-text'
                        value={article}
                        onChange={this.onChangeArticle}
                    />
                    <button
                        className='btn-article'
                        id='article-save'
                        onClick={this.update}
                        disabled={this.state.saveDisabled}
                    >
                        {this.state.saveText}
                    </button>
                </article>
                <Article
                    blogElement={this.state.blogElement}
                    backLink={true}
                    ellipsis={false}
                />
            </div>);
    }

    private readonly onChangeArticle = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
        event.preventDefault();
        const blogElement = { ...this.state.blogElement, article: event.target.value };
        this.setState({
            blogElement,
            saveText: this.SAVE,
            saveDisabled: false
        });
    }

    private readonly onChangeArticleTitle = (event: React.ChangeEvent<HTMLInputElement>) => {
        event.preventDefault();
        const blogElement = { ...this.state.blogElement, articleTitle: event.target.value };
        this.setState({
            blogElement,
            saveText: this.SAVE,
            saveDisabled: false
        });
    }

    private readonly update = (event: React.MouseEvent<HTMLButtonElement>) => {
        event.preventDefault();
        this.setState({
            saveText: this.SAVING,
            saveDisabled: true
        });
        const { article, articleTitle, blogDate, id } = this.state.blogElement;
        this.props.update(article, articleTitle, blogDate, id)
            .then(success => {
                this.setState({
                    saveText: success ? this.SAVED : this.FAILED
                });
            });
    }
}
