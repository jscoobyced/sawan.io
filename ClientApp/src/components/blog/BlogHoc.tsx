import React from 'react';
import { match } from 'react-router-dom';
import { ContentServiceFactory } from '../../services/ContentServiceFactory';
import { IContentService } from '../../services/IContentService';
import { BlogElement } from '../../services/Models';
import { AuthenticationFactory } from '../../utils/AuthenticationFactory';
import { DateUtil } from '../../utils/DateUtils';
import { IdParam } from '../Models';
import { Article } from './Article';
import { EditableArticle } from './EditableArticle';

export interface BlogHocProps {
    match: match<IdParam>;
}

export interface BlogState {
    blogElement: BlogElement;
}

export class BlogHoc extends React.Component<BlogHocProps, BlogState> {
    private readonly contentService: IContentService;

    public constructor(props: BlogHocProps, state: BlogState) {
        super(props, state);
        this.contentService = ContentServiceFactory.GetContentService();
        this.state = {
            blogElement: {
                article: '',
                articleTitle: '',
                blogDate: DateUtil.defaultDate(true),
                updateDate: DateUtil.defaultDate(true),
                id: 0
            }
        };
    }

    public componentDidUpdate(prevProps: BlogHocProps) {
        const articleId = +this.props.match.params.id;
        const prevArticleId = +prevProps.match.params.id;
        if (articleId !== prevArticleId) {
            this.doUpdate(articleId);
        }
    }

    public componentDidMount() {
        this.doUpdate(+this.props.match.params.id);
    }

    public updateBlog = (article: string, articleTitle: string, articleBlogDate: Date, id: number): Promise<boolean> => {
        const blogElement: BlogElement = {
            article,
            articleTitle,
            blogDate: articleBlogDate,
            updateDate: DateUtil.defaultDate(true),
            id
        };
        return this.contentService.saveBlogElement(blogElement);
    }

    public render() {
        const edit = this.props.match.params.edit === "edit";
        const isAdmin = AuthenticationFactory.getAuthentication() && AuthenticationFactory.getAuthentication().isAdmin();

        if (isAdmin && edit) {
            return (
                <EditableArticle
                    blogElement={this.state.blogElement}
                    update={this.updateBlog}
                />);
        } else {
            return (
                <Article
                    blogElement={this.state.blogElement}
                    backLink={true}
                    editLink={isAdmin}
                    ellipsis={false}
                />);
        }
    }

    private doUpdate(articleId: number) {
        if (articleId === 0) {
            return;
        }

        this.contentService.getBlogElement(articleId)
            .then(blogElement => {
                if (blogElement) {
                    this.setState({ blogElement });
                }
            });
    }
}
