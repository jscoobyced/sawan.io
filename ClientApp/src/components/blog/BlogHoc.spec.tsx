import { shallow } from 'enzyme';
import React from 'react';
import { match } from 'react-router-dom';
import { TestUtils } from '../../../tests/TestUtils';
import { ContentServiceFactory } from '../../services/ContentServiceFactory';
import { BlogElement } from '../../services/Models';
import { AuthenticationFactory } from '../../utils/AuthenticationFactory';
import { DateUtil } from '../../utils/DateUtils';
import { MockAuthentication } from '../auth/MockAuthentication';
import { IdParam } from '../Models';
import { Article } from './Article';
import { EditableArticle } from './EditableArticle';
import { BlogHoc } from './BlogHoc';

test('Blog is rendered empty when no blogElement', () => {
  const matchParam: match<IdParam> = TestUtils.createMatch('0', 'edit');
  const blog = shallow(<BlogHoc match={matchParam} />);
  expect(blog.find(Article)).toHaveLength(1);
  expect(blog.find(EditableArticle)).toHaveLength(0);
});

test('Blog is rendered with blogElement', async () => {
  const matchParam: match<IdParam> = TestUtils.createMatch('1');
  const blog = shallow(<BlogHoc match={matchParam} />);
  expect(blog.find(Article)).toHaveLength(1);
  expect(blog.find(EditableArticle)).toHaveLength(0);
});

test('Blog is rendered writable for admin', () => {
  const matchParam: match<IdParam> = TestUtils.createMatch('1', 'edit');
  AuthenticationFactory.registerAuthentication(new MockAuthentication());
  const blog = shallow(<BlogHoc match={matchParam} />);
  expect(blog.find(Article)).toHaveLength(0);
  expect(blog.find(EditableArticle)).toHaveLength(1);
});

test('Blog is rendered not writable for non-admin', () => {
  const mockAuthentication = new MockAuthentication();
  mockAuthentication.init = jest.fn().mockImplementation();
  mockAuthentication.isAdmin = jest.fn().mockImplementation(() => false);
  AuthenticationFactory.registerAuthentication(mockAuthentication, true);
  const matchParam: match<IdParam> = TestUtils.createMatch('1', 'edit');
  const blog = shallow(<BlogHoc match={matchParam} />);
  expect(blog.find(Article)).toHaveLength(1);
  expect(blog.find(EditableArticle)).toHaveLength(0);
});

test('Blog is rendered to new page', () => {
  const matchParam: match<IdParam> = TestUtils.createMatch('1');
  const blog = shallow(<BlogHoc match={matchParam} />);
  expect(blog.find(Article)).toHaveLength(1);
  const newMatchParam = TestUtils.createMatch('2');
  blog.setProps({ ...blog.props, match: newMatchParam });
  expect(blog.find(Article)).toHaveLength(1);
});

test('Blog is updated', async () => {
  const contentService = ContentServiceFactory.GetContentService();
  const { articleTitle } = await contentService.getBlogElement(1);
  expect(articleTitle).toBe('Starting again');
  const matchParam: match<IdParam> = TestUtils.createMatch('1');
  const blog = new BlogHoc({ match: matchParam }, { blogElement: null as unknown as BlogElement });
  blog.updateBlog('test', 'test', DateUtil.defaultDate(), 1);
  const articleTitleAfter = (await contentService.getBlogElement(1)).article;
  expect(articleTitleAfter).toBe('test');
});
