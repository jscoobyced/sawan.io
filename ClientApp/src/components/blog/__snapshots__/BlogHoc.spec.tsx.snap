// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog is rendered empty when no blogElement 1`] = `
<Article
  backLink={true}
  blogElement={
    Object {
      "article": "",
      "articleTitle": "",
      "blogDate": 2019-03-14T00:00:00.000Z,
      "id": "0",
      "updateDate": 2019-01-01T00:00:00.000Z,
    }
  }
  ellipsis={false}
/>
`;

exports[`Blog is rendered with blogElement 1`] = `
<Article
  backLink={true}
  blogElement={
    Object {
      "article": "",
      "articleTitle": "",
      "blogDate": 2019-03-14T00:00:00.000Z,
      "id": "0",
      "updateDate": 2019-01-01T00:00:00.000Z,
    }
  }
  ellipsis={false}
/>
`;

exports[`Blog is rendered writable 1`] = `
<EditableArticle
  blogElement={
    Object {
      "article": "",
      "articleTitle": "",
      "blogDate": 2019-03-14T00:00:00.000Z,
      "id": "0",
      "updateDate": 2019-01-01T00:00:00.000Z,
    }
  }
  update={[Function]}
/>
`;
